scipy/_lib/_array_api.py

```diff
@@ -163,3 +163,53 @@ def copy(x, *, xp=None):
         xp = array_namespace(x)
 
     return as_xparray(x, copy=True, xp=xp)
+
+
+def assert_equal(actual, desired, err_msg='', xp=None):
+    # TODO compare arrays from different libraries
+    if xp is None:
+        xp = array_namespace(actual)
+    if 'cupy' in xp.__name__:
+        return xp.testing.assert_array_equal(actual, desired, err_msg=err_msg)
+    elif 'torch' in xp.__name__:
+        return xp.testing.assert_close(actual, desired, rtol=0, atol=0,
+                                       msg=err_msg)
+    return np.testing.assert_array_equal(actual, desired, err_msg=err_msg)
+
+
+def assert_close(actual, desired, rtol=None, atol=None, err_msg='', xp=None):
+    # TODO compare arrays from different libraries
+    if xp is None:
+        xp = array_namespace(actual)
+    if 'cupy' in xp.__name__:
+        if rtol is None:
+            rtol = 1e-07
+        if atol is None:
+            atol = 0
+        return xp.testing.assert_allclose(actual, desired, rtol=rtol,
+                                          atol=atol, err_msg=err_msg)
+    elif 'torch' in xp.__name__:
+        return xp.testing.assert_close(actual, desired, rtol=rtol,
+                                       atol=atol, msg=err_msg)
+    if rtol is None:
+        rtol = 1e-07
+    if atol is None:
+        atol = 0
+    return np.testing.assert_allclose(actual, desired, rtol=rtol,
+                                      atol=atol, err_msg=err_msg)
+
+
+def assert_less(actual, desired, err_msg='', verbose=True, xp=None):
+    # TODO compare arrays from different libraries
```

Comment:
How does this gel with `_assert_matching_namespace` from: https://github.com/scipy/scipy/pull/19005/files#diff-b440e60b73eceb9148588107815b1b5cd66d6858e3f56f8d4872584b704a02e7R172 ?

Or maybe there's a concrete example in mind where we aren't ready to strictly require the namespace preservation?

Reply:
The motivation for this is to be able to, say, write `assert_equal(x, y)` where `x` is an `xp` array and `y` is an `ndarray`. This purely makes writing the tests easier. Our public functions should still avoid device transfers - this is just a convenience to make the process of modifying tests easier.

As for how it gels with `_assert_matching_namespace`, I don't see any issues. Couldn't we still use it immediately after the function we are testing? Or, as I have it in #19005, just in a separate class.

Reply:
I think I'd be happy to requite matching namespaces, unless there's a good reason why this won't suffice:

> Perhaps it is better to always call `xp.asarray` from within the test though

Reply:
Also noting that, if we do require the namespaces to match, `array_namespace(actual, desired)` should be used.

Reply:
> Is the idea that instead of converting from xp to np to perform the comparison with np.testing, we convert from np to xp and perform the comparison with the xp's testing function?

Yes.

> If that's right, I take it that xp.asarray is guaranteed to accept a np array as input (for any Array API compatible xp), even if np.asarray is not guaranteed to work on an xp array? (Otherwise, we'd just reversed the original problem.)

I think so. [`xp.asarray`](https://data-apis.org/array-api/latest/API_specification/generated/array_api.asarray.html) accepts any object which supports the Python buffer protocol. There isn't a clear answer from google but [this page](https://cython.readthedocs.io/en/latest/src/userguide/memoryviews.html) seems to suggest that `numpy.ndarray`s do support it.

> I think that separating the details of array backends from the intent of the test would be nice. As the test author, I would just want these functions to compare the data in the arrays (and possibly shape and dtype - something that I think is missing in some of the np.testing functions), regardless of backend.

Totally agree with this. Abstracting as much as possible out to these assertions seems desirable.

> If xp.asarray is guarranteed to work with np input, requiring the conversion in the test would suffice, but if it's going to be done in every test, it would be more convenient to just put it inside the function.

I agree. If Tyler and Ralf are happy with this then we can start to think about which library's assertions to use for each combination of arrays, and how to implement this.

Reply:
> I think so. xp.asarray accepts any object which supports the Python buffer protocol. 

Just learned that the buffer protocol is CPU only, so it seems not.

Reply:
> Just learned that the buffer protocol is CPU only, so it seems not.

It sounds like this reinforces the need for these assertions to do the conversions themselves, then.

Reply:
https://data-apis.org/array-api/2022.12/design_topics/data_interchange.html#data-interchange It looks like we should be using DLPack. Maybe we can build this into our `as_xparray` helper.

Reply:
I checked to see what scikit-learn are doing: https://github.com/scikit-learn/scikit-learn/blob/a05eb6bbe30bc40904274c87afb4ac93bcd60168/sklearn/model_selection/tests/test_split.py#L1286-L1335

They are also just using `xp.asarray` to go from `np` to `xp`. It's true that this is not guaranteed to work for any library, but it's sufficient while we only test with libraries for which it works. I don't know how likely it is that we will want to test with a library for which this doesn't work.

For `assert_allclose`, they are converting back to `np` using `_convert_to_numpy`, which has the necessary shims for when `np.asarray` doesn't work: https://github.com/scikit-learn/scikit-learn/blob/2a2772a87b6c772dc3b8292bcffb990ce27515a8/sklearn/utils/_array_api.py#L191-L204.

Reply:
tl;dr `xp.asarray(a_numpy_array)` should be robust enough, and about the best we can do.

More detailed: the buffer protocol is CPU-only indeed. So CuPy does something special to make it efficient, converting directly from NumPy. But worst-case, if a library does not support NumPy arrays directly )_and_ also not the buffer protocol, then a numpy array should still get treated like a Python list (or list of lists), and iterated over - extracting the data element by element. The `asarray` spec contains `NestedSequence` as an input, so that is supposed to work. There is no more generic option.

I can't think of a current real-world scenario where this wouldn't work. If a library doesn't support CPUs at all, it'd have much bigger problems of course, so I don't think we have to care about such hypothetical corner cases.

Reply:
> (and possibly shape and dtype - something that I think is missing in some of the `np.testing` functions)

+1. I just noticed this earlier when looking at Lucas's `fft` PR: PyTorch is more strict here, so a direct translation of tests from numpy to array API was failing here.